### PR TITLE
fix(@angular/cli): use webpack 4

### LINKED ISSUE #9058: Default import is undefined
### Versions
```
<!--
Output from: `ng --version`.
If nothing, output from: `node --version` and `npm --version`.
  Windows (7/8/10). Linux (incl. distribution). macOS (El Capitan? Sierra?)
-->
Angular CLI: 1.6.3
Node: 9.3.0
OS: darwin x64
Angular: 5.1.2
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.6.3
@angular-devkit/build-optimizer: 0.0.36
@angular-devkit/core: 0.0.22
@angular-devkit/schematics: 0.0.42
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.9.3
@schematics/angular: 0.1.11
@schematics/schematics: 0.0.11
typescript: 2.5.3
webpack: 3.10.0
```

### Repro steps

* git clone https://github.com/cebor/angular-highcharts-demo
* ng serve
* See Errors in Console

### Observed behavior
```
<!-- Normally this includes a stack trace and some more information. -->
TypeError: chartModule is not a function
    at eval (webpack-internal:///../../../../angular-highcharts/angular-highcharts.es5.js:257)
    at Array.forEach (<anonymous>)
    at ChartService.initModules (webpack-internal:///../../../../angular-highcharts/angular-highcharts.es5.js:256)
    at new ChartModule (webpack-internal:///../../../../angular-highcharts/angular-highcharts.es5.js:286)
    at _createClass (webpack-internal:///../../../core/esm5/core.js:11112)
    at _createProviderInstance$1 (webpack-internal:///../../../core/esm5/core.js:11086)
    at initNgModule (webpack-internal:///../../../core/esm5/core.js:11039)
    at new NgModuleRef_ (webpack-internal:///../../../core/esm5/core.js:12293)
    at createNgModuleRef (webpack-internal:///../../../core/esm5/core.js:12283)
    at Object.debugCreateNgModuleRef [as createNgModuleRef] (webpack-internal:///../../../core/esm5/core.js:14627)
```

### Desired behavior

Following default import should not be undefined: https://github.com/cebor/angular-highcharts-demo/blob/master/src/app/app.module.ts#L4

### Mention any other details that might be useful (optional)
* ng serve --prod is working
* with ng 1.6.1 all works fine

### LINKED ISSUE #9069: bug: all RxJS operators pulled into prod bundle when using lettable imports from 'rxjs/operators'
### Versions
```
Angular CLI: 1.6.3
Node: 8.9.0
OS: darwin x64
Angular: 5.1.2
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

@angular/cdk: 5.0.2
@angular/cli: 1.6.3
@angular/flex-layout: 2.0.0-beta.12
@angular/material: 5.0.2
@angular-devkit/build-optimizer: 0.0.36
@angular-devkit/core: 0.0.22
@angular-devkit/schematics: 0.0.42
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.9.3
@schematics/angular: 0.1.11
@schematics/schematics: 0.0.11
typescript: 2.5.3
webpack: 3.10.0
```

### Repro steps
https://github.com/Toxicable/issue-lettable-bundle-size

### Observed behavior
From https://github.com/angular/angular-cli/issues/8854#issuecomment-354532989, "There appears to be a bug in Webpack that is causing the CLI to import all of the RXJS operators if you use any lettable operator imports without doing each as a deep import individually.

Upgrading to lettable operators in CLI 1.6.3 caused RxJS to go from 60 KB to 122 KB for me. It sounds like a fix is 'coming soon', but I haven't been able to track down the issue to follow."

Another user mentioned a bundle size increase when moving to lettable operators in the CLI: https://github.com/angular/angular-cli/issues/8720#issuecomment-348804558

### Desired behavior
From the [5.0 blog post](https://blog.angular.io/version-5-0-0-of-angular-now-available-37e414935ced):
"These new operators eliminate the side effects and the code splitting / tree shaking problems that existed with the previous ‘patch’ method of importing operators."
"RxJS now distributes a version using ECMAScript Modules. The new Angular CLI will pull in this version by default, **saving considerably on bundle size**."

I desire a considerable reduction in bundle size, not an increase.

### Mention any other details that might be useful (optional)
I tracked this down across Slacks, GitHub, and [Twitter](https://twitter.com/Splaktar/status/948039571681538050) as I was told that there was a known Webpack problem that would be fixed soon and then pulled into the CLI. As far as I can tell, any similar Webpack issue was solved in `3.9.0` and the current CLI already includes it as it is using `3.10.0` now.

I can provide source map explorer screenshots if needed, but the linked repo above should enable you to better reproduce and create your own source map explorer output.

RxJS Docs for lettable operators and builds: https://github.com/ReactiveX/rxjs/blob/master/doc/lettable-operators.md#build-and-treeshaking

#### Workaround
Breaking out all of your imports into individual (one import per line) deep imports seems to avoid this issue, but it is very painful and not desirable to have these extra lines of imports in projects of any significant size or scope.

### LINKED ISSUE #9135: "ng serve --prod --build-optimizer" live-reloads cause Javascript error in browsers
### Versions
```
Angular CLI: 1.6.3
Node: 8.9.3
OS: win32 x64
Angular: 5.1.3
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.6.3
@angular-devkit/build-optimizer: 0.0.36
@angular-devkit/core: 0.0.22
@angular-devkit/schematics: 0.0.42
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.9.3
@schematics/angular: 0.1.11
@schematics/schematics: 0.0.11
typescript: 2.4.2
webpack: 3.10.0
```

### Repro steps
* `ng new repro-app`
* `ng serve --prod --build-optimizer`
* Add css to styles.css in VS Code and save. 
* While the project is rebuilding, also change and save index.html.
* Inspect browser console for error output
* Subsequent rebuilds still throw the error even if only changing one file at a time. 

### Observed behavior
The first build is served with no issue. The third build generates an error, which is below. 
**CLI Output**
```
 ng serve --prod --build-optimizer
****************************************************************************************
This is a simple server for use in testing or debugging Angular applications locally.
It hasn't been reviewed for security issues.

DON'T USE IT FOR PRODUCTION!
****************************************************************************************
** NG Live Development Server is listening on localhost:4200, open your browser on http://localhost:4200/ **
Date: 2018-01-08T21:34:33.896Z
Hash: dcf6fdafe5526d65c989
Time: 19290ms
chunk {0} polyfills.61df7d7ec492d95bb0b2.bundle.js (polyfills) 61.2 kB [initial] [rendered]
chunk {1} main.da0014a7afa74a009599.bundle.js (main) 293 kB [initial] [rendered]
chunk {2} styles.7603117359b5a8dffd89.bundle.css (styles) 105 bytes [initial] [rendered]
chunk {3} inline.d3c0cf3dc365c4d3fdfc.bundle.js (inline) 1.45 kB [entry] [rendered]

webpack: Compiled successfully.
webpack: Compiling...
Date: 2018-01-08T21:34:55.521Z
Hash: e1d88d3cedb984919ba1
Time: 13195ms
chunk {0} polyfills.61df7d7ec492d95bb0b2.bundle.js (polyfills) 61.3 kB [initial]
chunk {1} main.da0014a7afa74a009599.bundle.js (main) 293 kB [initial]
chunk {2} styles.8b89660c6d7db7fd0150.bundle.css (styles) 106 bytes [initial]
chunk {3} inline.d3c0cf3dc365c4d3fdfc.bundle.js (inline) 1.45 kB [entry]

webpack: Compiled successfully.
webpack: Compiling...
webpack: wait until bundle finished: /dashboard
webpack: wait until bundle finished: /dashboard
Date: 2018-01-08T21:35:08.411Z
Hash: 765ff0a403db99858e58
Time: 12637ms
chunk {0} polyfills.61df7d7ec492d95bb0b2.bundle.js (polyfills) 61.3 kB [initial]
chunk {1} main.29d5d263a9d4e91c64ff.bundle.js (main) 280 kB [initial] [rendered]
chunk {2} styles.8b89660c6d7db7fd0150.bundle.css (styles) 106 bytes [initial]
chunk {3} inline.c6299d98af44410d31e7.bundle.js (inline) 1.45 kB [entry] [rendered]

webpack: Compiled successfully.
```
**Browser Console Stack Trace**
```
Uncaught TypeError: (void 0) is not a function
    at Object.cDNt (main.2d2b3f0de20dee881539.bundle.js:1)
    at o (inline.a443620b1a7e83c745b9.bundle.js:1)
    at Object.0 (main.2d2b3f0de20dee881539.bundle.js:1)
    at o (inline.a443620b1a7e83c745b9.bundle.js:1)
    at window.webpackJsonp (inline.a443620b1a7e83c745b9.bundle.js:1)
    at main.2d2b3f0de20dee881539.bundle.js:1
cDNt @ main.2d2b3f0de20dee881539.bundle.js:1
o @ inline.a443620b1a7e83c745b9.bundle.js:1
0 @ main.2d2b3f0de20dee881539.bundle.js:1
o @ inline.a443620b1a7e83c745b9.bundle.js:1
window.webpackJsonp @ inline.a443620b1a7e83c745b9.bundle.js:1
(anonymous) @ main.2d2b3f0de20dee881539.bundle.js:1
```

### Desired behavior
The rebuilt webapp should be served on every rebuild.

### PR DESCRIPTION
**Pull request by filipesilva**
Just trying it out for now. Latest  version changelog and feedback thread at https://github.com/webpack/webpack/issues/6357.

Webpack slated for 4.0 final in 25th of February (see https://medium.com/webpack/webpack-4-beta-try-it-today-6b1d27d7d7e2).

To-dos:
- [ ] use `webpack.debug.ProfilingPlugin()`
- [ ] configure `System.import` usage (see https://github.com/webpack/webpack/pull/6321)
- [ ] replace `CommonsChunkPlugin` (see https://github.com/webpack/webpack/pull/6341 and https://gist.github.com/sokra/1522d586b8e5c0f5072d7565c2bee693)

Blockers:
- [x] https://github.com/webpack-contrib/copy-webpack-plugin/pull/174 (not an issue with alpha.0)
- [ ] https://github.com/jantimon/html-webpack-plugin/issues/816
- [x] postcss-loader uses loader._compilation.applyPluginsWaterfall (2.x doesn't)
- [ ] file-loader uses options.context
- [ ] rxjs path mapping breaks https://github.com/webpack/webpack/issues/6179#issuecomment-355296412

Warnings:
- [ ] `copy-webpack-plugin` 
  - `Tapable.plugin is deprecated. Use new API on `.hooks` instead`
- [ ] `extract-text-webpack-plugin`
  - `DeprecationWarning: Chunk.entrypoints: Use Chunk.getEntrypoints()`
  - `Chunk.entrypoints: Use Chunk.addEntrypoint/setEntrypoints to modify entrypoints.`
  - `Chunk.chunks: Use Chunk.getChunks() instead`
  - `Chunk.parents: Use Chunk.getParents() instead`
- [ ] `./src/paths-plugin.ts`
  - `DeprecationWarning: The callback argument was splitted into resolveContext and callback.`
- [ ] `./angular_compiler_plugin.ts`
  - `DeprecationWarning: Tapable.apply is deprecated. Call apply on the plugin directly instead`
- [ ] `./src/scripts-webpack-plugin.ts`
  - `Tapable.plugin is deprecated. Use new API on '.hooks' instead`
- [ ] `WARNING in asset size limit` for large bundles/assets
  
Fix #9058
Fix #9069
Fix #9135

### CODE REVIEW

File: packages/@angular/cli/models/webpack-configs/browser.ts [hunk 1]
```diff
@@ -88,18 +88,19 @@ export function getBrowserConfig(wco: WebpackConfigOptions) {
...
-        excludeChunks: lazyChunks,
-        xhtml: true,
-        minify: buildOptions.target === 'production' ? {
-          caseSensitive: true,
-          collapseWhitespace: true,
-          keepClosingSlash: true
-        } : false
-      }),
+      // Not yet supported in webpack 4.
+      // new HtmlWebpackPlugin({
```
Comment (pleerock) [inline packages/@angular/cli/models/webpack-configs/browser.ts]: I guess I do not have `index.html` in build because of the following code?

Comment (filipesilva) [reply packages/@angular/cli/models/webpack-configs/browser.ts]: Correct, we're still waiting for plugins to be updated. At the moment this PR is not ready.

Comment (pleerock) [reply packages/@angular/cli/models/webpack-configs/browser.ts]: Thank you, looks like we are expecting #816 to be resolved

**Comment by filipesilva** [comment]
Superseded by https://github.com/angular/angular-cli/pull/9459.

**Pull request closed without merging**